### ISSUE TITLE
Datetime support

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#1097]
TL;DR

- Use `int64_t` with unit (`ns`) as *duration*.
- Add dtype `std::chrono::time_point` (dtype name `datetime64`) using `ns` duration and system(?) clock
- Time points *should* probably have a unit for convenience (unlike mentioned in chat log below).
- Implement operators, e.g., += between a scipp.datetime64 and scipp.int64. The operation would check that the unit of the integer operand ~is ns~ is the same as that of the datetime, and the integers are converted to std::chrono::duration<int64_t, std::nano> on the fly.
- Figure out convenient ways for displaying readable strings.
- Python bindings, should be made compatible with `numpy.datetime64[ns]`.

Details from Slack:

06:56
Simon Heybrock Looking into ways for better handling of date-time objects like pulse-times. We would like to play well with numpy.datetime64 (which actually also supports different time units). boost::date_time doesn't look good (using 96 bits with nanosecond resolution), but since C++11 there is std::chrono (with more additions in C++20 to come). std::chrono::duration is essentially just an integer with a compile-time defined time unit such as ns. (edited) 
06:58
Simon Heybrock Since we have units in scipp we may not actually want duration as the type we store, we could just continue using integers, but it is not so clear to me right now. Do we use a time library only for display purposes?

2 replies
Last reply 20 days agoView thread
06:59
Simon Heybrock The second component is std::chrono::time_point, which is just a duration alongside a (compile-time) defined clock.
07:02
numpy.datetime64 "corresponds" to std::chrono::time_point and numpy.timedelta64 corresponds to std::chrono::duration.
07:03
Durations or time deltas are supported using normal scipp integer types with time units (the missing part would be support for, e.g., operations between ns and s, but it is not clear whether we want that or should require explicit conversions).
07:06
Time points (datetime) are not so clear. They should not just be scipp integers with, e.g. ns as a unit, since in principle we should not support additions of two time points, or multiplications with dimensionless integers (for durations this is ok: 100 ns * 10, but not for time points: 1st of February * 2). (edited) 
07:08
That is, we would probably want to introduce a datetime64 dtype, implemented as std::chrono::time_point. This would not have a scipp unit, but the time_point knows its resolution at compile time.
07:12
We can then implement, e.g., += between a scipp.datetime64 and scipp.int64. The operation would check that the unit of the integer operand is ns, and the integers are converted to std::chrono::duration<int64_t, std::nano> on the fly.
07:15
What remains is to figure out whether std::chrono::time_point is compatible with numpy.datetime64[ns] and pybind11 can deal with them, i.e., whether our Python bindings will be simple or a nightmare.
07:19
Actually, someone just made a PR for that: https://github.com/pybind/pybind11/issues/2078. (edited)

**Comment by rozyczko** [comment]
from slack discussion with Simon:

SH: We may want compatibility (but being compatible with numpy datetime should be sufficient). As explained in the issue, the scipp unit system should give us most of what we need. I think the task should be done in the following steps:

- Add new dtype (for time point) and operations with integers (used as time differences).
- Add Python bindings.
I expect 2.) to be the bigger challenge, but it might not.

PR: how about support for things like resampling?
SH: I would worry about this later.

SH: we might need to add `ns` as a unit

**Comment by rozyczko** [comment]
The pybind patch casts `np.datetime64` values into `datetime.datetime`, which has only microsecond precision. This effectively cuts nanoseconds out.
Passing `np.datetime64('2011-06-15T12:00:00.123456')` returns correct `2011-06-15T12:00:00.123456` as `datetime`.
Passing `np.datetime64('2011-06-15T12:00:00.1234567')` fails with the patched pybind.

Possible solution 1:
In the pybind11 patch - retrieve the explicit nanoseconds with e.g. `(np_time -np.t.astype('datetime64[us]')).astype(int)` and then add it to the resulting `system_clock::from_time_t` like it's done with miliseconds. This requires removing the nanosecond part from the passed np.datetype64 value in order to have it properly cast into `datetime.datetime`.

Possible solution 2:
Don't cast `np.datetime64` into `datetime.datetime` as in the patch, but rather retrieve all the components (from year through nanosecond) and create `std::tm` to use as argument to `std::mktime`.

**Comment by rozyczko** [comment]
Actually, this is slightly more sinister.
`chrono.h` in pybind11 is very `datetime.datetime` centric with several places going explicitly down to microsecond granularity.
The casting function is using `std::micro` for duration accuracy, again, casting into `datetime.datetime`, so even if the accuracy is changed to `std::nano`.
Maybe we can rewrite pybind11's chrono.h to not use datetime.datetime?

**Comment by rozyczko** [comment]
I made a proof of concept patch for `chrono.h` in pybind11, accepting `numpy.datetime64`, converting internally to `chrono::system_clock::time_point` and converting back. This uses timestamps and needs more work, especially wrt assuring current behaviour is correct (converting non-numpy datetime to `time_point` through `std::tm` and `std::localtime` and back to `datetime.datetime`)

**Comment by rozyczko** [comment]
Added `-` to allow creating duration between two time points.
Note that `+` for two time points is undefined ( https://en.cppreference.com/w/cpp/chrono/time_point/operator_arith2 )

**Comment by rozyczko** [comment]
Applying the patch for `chrono.h` in pybind11 is reasonably easy with CMake's `add_custom_command` but I wonder if we should implement it, considering we might want to switch to conda pulls for dependencies ( #1241 ).
If we do change, patching the installed package will be slightly more difficult but not impossible.
I think we should consider local patching given history of `pybind11` pull requests sitting unmerged for long time...

**Comment by SimonHeybrock** [comment]
I think I agree, and this is another reason not to switch to use conda.

**Comment by SimonHeybrock** [comment]
Binding the chrono stuff may change significantly relating to https://github.com/scipp/scipp/pull/1225#discussion_r458613461.

**Comment by rozyczko** [comment]
Added a new `to_string` function allowing printing of properly formatted time points.
For `ns`, we get datetime represented as `2020-07-27T10:41:11.200000011`
for `s` we get `2020-07-27T10:41:11`
